src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -624,42 +619,81 @@ ucs_status_t uct_dc_mlx5_ep_flush(uct_ep_h tl_ep, unsigned flags,
+        return UCS_INPROGRESS;
+    }
 
-        if (uct_dc_mlx5_iface_is_dci_rand(iface)) {
-            return UCS_ERR_UNSUPPORTED;
-        }
+    status = uct_ib_mlx5_modify_qp_state(md, &txwq->super, IBV_QPS_ERR);
+    if (status != UCS_OK) {
+        return status;
+    }
 
-        if (ep->flags & UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL) {
-            goto out;
-        }
+    uct_ib_mlx5_txwq_update_flags(txwq, UCT_IB_MLX5_TXWQ_FLAG_FAILED, 0);
+    ep->flags |= UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL;
+    *sn_p      = txwq->sw_pi;
 
-        status = uct_ib_mlx5_modify_qp_state(md, &txwq->super, IBV_QPS_ERR);
-        if (status != UCS_OK) {
+    /* Post NOP on the DCI to avoid releasing it from dci_put while it's in
+     * error state. We could check its state when it's released, but it would
+     * add another branch on the fast path. */ 
+    uct_rc_mlx5_txqp_inline_post(&iface->super, UCT_IB_QPT_DCI, txqp, txwq,
+                                 MLX5_OPCODE_NOP, NULL, 0, 0, 0, 0, 0, 0,
+                                 &ep->av, uct_dc_mlx5_ep_get_grh(ep),
+                                 uct_ib_mlx5_wqe_av_size(&ep->av), 0, INT_MAX);
+
+    return UCS_INPROGRESS;
+}
+
+ucs_status_t uct_dc_mlx5_ep_flush(uct_ep_h tl_ep, unsigned flags,
+                                  uct_completion_t *comp)
+{
+    uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface,
+                                                uct_dc_mlx5_iface_t);
+    uct_dc_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
+    uint8_t pool_index         = uct_dc_mlx5_ep_pool_index(ep);
+    ucs_status_t status;
+    uint16_t sn;
+    UCT_DC_MLX5_TXQP_DECL(txqp, txwq);
+
+    if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {
+        status = uct_dc_mlx5_ep_flush_cancel(ep, &sn);
+        if (status != UCS_INPROGRESS) {
             return status;
         }
 
-        ep->flags |= UCT_DC_MLX5_EP_FLAG_FLUSH_CANCEL;
-        sn         = txwq->sw_pi;
-        /* post NOP operation which will complete with error, to trigger DCI
-         * reset. Otherwise, DCI could be returned to poll in error state */
-        uct_rc_mlx5_txqp_inline_post(&iface->super, UCT_IB_QPT_DCI,
-                                     txqp, txwq,
-                                     MLX5_OPCODE_NOP, NULL, 0,
-                                     0, 0, 0,
-                                     0, 0,
-                                     &ep->av, uct_dc_mlx5_ep_get_grh(ep),
-                                     uct_ib_mlx5_wqe_av_size(&ep->av),
-                                     0, INT_MAX);
+        goto out;
+    }
+
+    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
+        if (uct_dc_mlx5_iface_dci_can_alloc(iface, pool_index)) {
+            UCT_TL_EP_STAT_FLUSH(&ep->super);
+            return UCS_OK; /* nothing to flush */
+        }
+
+        return UCS_ERR_NO_RESOURCE; /* waiting for dci */
+    }
+
+    if (!uct_dc_mlx5_iface_has_tx_resources(iface) ||
```

Comment:
maybe uct_dc_mlx5_iface_has_tx_resources check could be done before `if (ucs_unlikely(flags & UCT_FLUSH_FLAG_CANCEL)) {`

Reply:
as @brminich correctly noted, there is not reason to check it for flush(CANCEL)

Reply:
right :)